logstreams/src/main/java/io/zeebe/logstreams/impl/log/fs/FsLogSegments.java

```diff
@@ -7,80 +7,77 @@
  */
 package io.zeebe.logstreams.impl.log.fs;
 
-public class FsLogSegments {
-  protected int initialSegmentId = -1;
+import java.util.concurrent.CopyOnWriteArrayList;
 
-  protected FsLogSegment[] segments = new FsLogSegment[0];
+public final class FsLogSegments {
 
-  protected volatile int segmentCount = 0;
+  protected CopyOnWriteArrayList<FsLogSegment> segments;
 
-  public void init(int initialSegmentId, FsLogSegment[] initialSegments) {
-    this.segments = initialSegments;
-    this.initialSegmentId = initialSegmentId;
-    this.segmentCount = initialSegments.length; // volatile store
+  private FsLogSegments(FsLogSegment[] initialSegments) {
+    this.segments = new CopyOnWriteArrayList<>(initialSegments);
+  }
+
+  public static FsLogSegments fromFsLogSegmentsArray(FsLogSegment[] initialSegments) {
+    return new FsLogSegments(initialSegments);
   }
 
   /** invoked by the conductor after a new segment has been allocated */
   public void addSegment(FsLogSegment segment) {
-    final FsLogSegment[] newSegments = new FsLogSegment[segments.length + 1];
-
-    System.arraycopy(segments, 0, newSegments, 0, segments.length);
-    newSegments[segments.length] = segment;
-    this.segments = newSegments;
-
-    this.segmentCount = newSegments.length; // volatile store
+    segments.add(segment);
   }
 
-  public void removeSegmentsUntil(int segmentId) {
-    final int segmentIdx = segmentId - initialSegmentId;
-    final int newLength = segments.length - segmentIdx;
-    final FsLogSegment[] newSegments = new FsLogSegment[newLength];
+  public FsLogSegment getSegment(int segmentId) {
+    final int firstSegmentId = getFirstSegmentId();
+    final int segmentIdx = segmentId - firstSegmentId;
 
-    System.arraycopy(segments, segmentIdx, newSegments, 0, newLength);
-    this.segments = newSegments;
-    initialSegmentId += segmentIdx;
-    this.segmentCount = newSegments.length; // volatile store
+    if (0 <= segmentIdx && segmentIdx < segments.size()) {
+      return segments.get(segmentIdx);
+    } else {
+      return null;
+    }
   }
 
-  public FsLogSegment getSegment(int segmentId) {
-    final int segmentCount = this.segmentCount; // volatile load
+  public void deleteSegmentsUntil(int segmentId) {
+    // we need to remove the segment from the collection as first
+    // before closing it - otherwise other threads will see the segment still in the
+    // collection when it is already closed (for example if we use #removeIf)
+    // Note: Iterator#remove is not implemented by CopyOnWriteArrayList
 
-    final FsLogSegment[] segments = this.segments;
+    int firstSegmentId = getFirstSegmentId();
+    while (firstSegmentId != segmentId) {
```

Comment:
What if `segmentId < firstSegmentId`? I think we delete everything then, no?

Reply:
Yes you're right I rewrote this multiple times where then somethings get lost :sweat_smile: 
I will not expect that this case happened but I will add an check